sdks/java/core/src/main/java/org/apache/beam/sdk/io/WriteFiles.java

```diff
@@ -410,13 +412,44 @@ private GatherResults(Coder<ResultT> resultCoder) {
       } else {
         // Pass results via a side input rather than reshuffle, because we need to get an empty
         // iterable to finalize if there are no results.
-        return input
-            .getPipeline()
-            .apply(Reify.viewInGlobalWindow(input.apply(View.asList()), ListCoder.of(resultCoder)));
+        return input.apply("ToList", Combine.globally(new ToListCombineFn<>()));
```

Comment:
Seems like this was applied to all file-based IO not just TFRecordIO as mentioned in the PR description. Have we done enough experiments to make sure this won't have unintended adverse performance consequences in batch/streaming ? For example, what if the list does not fit in memory ?

Apologies if this was already discussed in the dev list.

cc: @robertwb @lukecwik @iemejia

Reply:
I believe this could be a regression based upon the above comment a few lines higher then where the code was changed:
```
// Pass results via a side input rather than reshuffle, because we need to get an empty
// iterable to finalize if there are no results.
```
The combine globally in the global window will produce a default output in the global window but will throw an exception if used against non-global window PCollections.

Performance wise this is likely an improvement overall since runners typically do a better job with GBK then with side inputs.

Reply:
Thanks. Agree with Luke. Combine globally has shuffle (GBK) inside hence breaks above statements. This could be a regression at least for Dataflow when there are no outputs. We should try following cases.

(1) Dataflow (and possibly other runners may have similar regressions ?) with an empty output. 
(2) Writing using a non-global window while WriteFiles.withWindowedWrites() not set.

Luke, does that make sense ? Anything else to try out to make sure there's no regression here ?

Reply:
What's the verdict here ? I suggest we revert this and introduce as an optional change since this is a potential regression for Dataflow runner.

Reply:
Created https://issues.apache.org/jira/browse/BEAM-9734 to make sure that this does not get into 2.21.0

Reply:
Imo just a revert would be better - giving an option to something that is not validated and can have unexpected adverse effects would result in user's confusion

Reply:
I'm not sure what's the verdict right now. I think we should revert this and discuss further steps with repo free of regressions. Tagging @lukecwik for input about leaving the optimization for global windows. @chamikaramj how do you think the optional change could be applied? A runtime flag for the runner?

Reply:
You could always apply the trigger if it is guaranteed to produce output by checking what the trigger definition is. I don't think you need a flag.

I'm not sure how much of a holistic picture of the whole WriteFiles transform was done to see if there is a better way to structure the transforms.

Reply:
Thanks for reverting.

My suggestion was to fork the transform expansion to use GlobalCombine when specifically requested through a transform parameter (for example see how we fork TextIO.Read expansion based on withHintMatchesManyFiles()).

https://github.com/apache/beam/blob/master/sdks/java/core/src/main/java/org/apache/beam/sdk/io/TextIO.java#L371
https://github.com/apache/beam/blob/master/sdks/java/core/src/main/java/org/apache/beam/sdk/io/TextIO.java#L401

I agree that the more flags we add the more confusing it becomes for users. So we should make sure that there are significant advantages in adding this option through performance benchmarking.